msbuild/Xamarin.MacDev.Tasks.Core/Tasks/CoreMLCompilerTaskBase.cs

```diff
@@ -0,0 +1,239 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using System.Diagnostics;
+using System.Collections;
+using System.Collections.Generic;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.MacDev.Tasks
+{
+	public abstract class CoreMLCompilerTaskBase : Task
+	{
+		string toolExe;
+
+		public string ToolName { get { return "coremlc"; } }
+
+		#region Inputs
+
+		public string SessionId { get; set; }
+
+		public bool EnableOnDemandResources { get; set; }
+
+		[Required]
+		public string IntermediateOutputPath { get; set; }
+
+		[Required]
+		public ITaskItem[] Models { get; set; }
+
+		[Required]
+		public string ProjectDir { get; set; }
+
+		[Required]
+		public string ResourcePrefix { get; set; }
+
+		string sdkDevPath;
+		public string SdkDevPath {
+			get { return string.IsNullOrEmpty (sdkDevPath) ? "/" : sdkDevPath; }
+			set { sdkDevPath = value; }
+		}
+
+		public string ToolExe {
+			get { return toolExe ?? ToolName; }
+			set { toolExe = value; }
+		}
+
+		public string ToolPath { get; set; }
+
+		#endregion
+
+		#region Outputs
+
+		[Output]
+		public ITaskItem[] BundleResources { get; set; }
+
+		[Output]
+		public ITaskItem[] PartialAppManifests { get; set; }
+
+		#endregion
+
+		string DeveloperRootBinDir {
+			get { return Path.Combine (SdkDevPath, "usr", "bin"); }
+		}
+
+		string GetFullPathToTool ()
+		{
+			if (!string.IsNullOrEmpty (ToolPath))
+				return Path.Combine (ToolPath, ToolExe);
+
+			var path = Path.Combine (DeveloperRootBinDir, ToolExe);
+
+			return File.Exists (path) ? path : ToolExe;
+		}
+
+		static ProcessStartInfo GetProcessStartInfo (IDictionary<string, string> environment, string tool, string args)
+		{
+			var startInfo = new ProcessStartInfo (tool, args);
+
+			startInfo.WorkingDirectory = Environment.CurrentDirectory;
+
+			foreach (var variable in environment)
+				startInfo.EnvironmentVariables[variable.Key] = variable.Value;
+
+			startInfo.CreateNoWindow = true;
+
+			return startInfo;
+		}
+
+		int Compile (ITaskItem item, string outputDir, string log, string partialPlist)
+		{
+			var environment = new Dictionary<string, string> ();
+			var args = new ProcessArgumentBuilder ();
+
+			args.Add ("compile");
+			args.AddQuoted (item.ItemSpec);
+			args.AddQuoted (Path.GetFullPath (outputDir));
+			args.Add ("--output-partial-info-plist");
+			args.AddQuoted (partialPlist);
+
+			var startInfo = GetProcessStartInfo (environment, GetFullPathToTool (), args.ToString ());
+			var errors = new StringBuilder ();
+			int exitCode;
+
+			try {
+				Log.LogMessage (MessageImportance.Normal, "Tool {0} execution started with arguments: {1}", startInfo.FileName, startInfo.Arguments);
+
+				using (var stdout = File.CreateText (log)) {
+					using (var stderr = new StringWriter (errors)) {
+						var process = ProcessUtils.StartProcess (startInfo, stdout, stderr);
+
+						process.Wait ();
+
+						exitCode = process.Result;
```

Comment:
System.Diagnostics.Process is disposable, you can dispose `process` here (or put it in a `using` clause).

Reply:
Ah, hmmm, that probably means other build tasks need the same change since this is largely c&p'd from I think the IBTool task (base class).

Reply:
It's not a necessary change, the current code would probably work fine, but iirc each process object keeps file descriptors open, and if it takes a while until the GC runs you might run into problems with the amount of file descriptors (and the msbuild tasks usually creates a lot of subprocesses).

Reply:
I think it's worth fixing then...